working/enhanced-type-promotion/feature-specification.md

```diff
@@ -0,0 +1,630 @@
+complex in some cases. But, in practice, the current rules are too restrictive.
+For example, this does not work:
+
+```dart
+int stringLength(Object o) {
+  if (o is! String) {
+    throw "Not a string.";
+  } else {
+    return o.length; // Error: Object does not have a "length" getter.
+  }
+}
+```
+
+The type promotion rules don't handle negated `is` checks and else clauses. This
+doesn't work either:
+
+```dart
+int stringLength(Object o) {
+  if (o is! String) throw "Not a string.";
+  return o.length; // Error: Object does not have a "length" getter.
+}
+```
+
+Since the first line will definitely throw if `o` is not a String, the only way
+the second line could execute is if `o` is a string. But type promotion fails to
+understand that.
+
+To get around these limitations, the analyzer used by IntelliJ and others also
+has "type propagation". This is more sophisticated than type promotion. But,
+since the language spec controls what errors and warnings are shown, type
+propagation is limited to be using for things like auto-complete suggestions.
+
+Maintaining two similar but different flow typing systems adds complexity to our
+tools and confuses users. ("Why are you showing me an error that a variable
+isn't a Foo, when auto-complete on that same variable shows me all of Foo's
+methods?")
+
+To remedy that, this proposal augments type promotion with smarter rules when
+variables can be promoted. That should avoid requiring annoying redundant casts
+in more places, and let us unify type promotion and propagation into a single
+coherent system. This proposal doesn't cover all possible extensions to type
+propagation, but it handles the simplest, most common cases.
+
+## How Type Promotion Works
+
+The language specifies type promotion out of four components:
+
+1.  **Certain expressions cause a fact to be known about a variable.** For
+    example, `o` is String deduces the fact "`o` must be a String".
+
+2.  **The scope where a fact is known to be true.** For example in:
+
+    ```dart
+    if (o is String) {
+      return o.length;
+    } else {
+      throw "Not a string.";
+    }
+    ```
+
+    The fact "`o` is a String" is known to be true inside the then clause of the
+    if statement, but not the else clause, and not after the end of the if
+    statement.
+
+3.  **Which variables can have their types promoted.** Some code may cause a
+    fact to not be soundly reliable. In that case, we shouldn't promote. For
+    example:
+
+    ```dart
+    test(Object o, bool callClosure) {
+      closure() {
+        o = 123;
+      }
+
+      if (o is String) {
+        if (callClosure) closure();
+        print(o.length);
+      }
+    }
+    ```
+
+    At the point where `.length` is accessed, `o` may be a String, or it may be
+    an int if `callClosure` was true. To avoid cases like this, the spec defines
+    certain variables to be off limits for type promotion.
```

Comment:
It would be great to elaborate a bit here. Under what conditions do we consider  a variable safe to type-promote?
```dart
void foo(x) async {
  closure() {
    x = 123;
  }

  if (x is String) {
    print(x.length); // Safe
  }

  if (x is String) {
    if (something) closure();
    print(x.length); // Unsafe
  }

  if (x is String) {
    await something;
    print(x.length); // Unsafe
  }
}
```

Reply:
Another example, a bit contrived:
```dart
var x = 'abc';

class A {
  bool get foo {
    x = 123;
    return true;
  }
}

void main() {
  var a = A();
  if (x is String && a.foo) {
    print(x.length); // Unsafe!
  }
}
```

Reply:
Yeah, these are good points. This proposal isn't quite active right now and I'm not the author, so I don't think we'll tweak it right now. But if we summon it back to life, I think we should be more explicit about the things you bring up.

Reply:
Line 120 says that the definition of which variables are safe to type promote is unchanged by this proposal, so it's what the spec says:

> The static type system ascribes a static type to every expression. In some cases, the type of a local variable (which can be a formal parameter) may be promoted from the declared type, based on control flow.

With the following caveat added in many places:

> ...  then the type of v is known to be T in s2, unless any of the following are true
> - v is potentially mutated in s1,
> - v is potentially mutated within a function other than the one where v is declared, or
> - v is accessed by a function defined in s1 and v is potentially mutated anywhere in the scope of v.